### DISCUSSION TITLE
Best way to access DB outside requests

### DISCUSSION THREAD

**Discussion by farahats9** [fastapi/fastapi#8595] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from typing import Generator
from sqlalchemy import create_engine
from sqlalchemy.orm import sessionmaker
from app.core.config import settings

engine = create_engine(settings.SQLALCHEMY_DATABASE_URI, pool_pre_ping=True, max_overflow=-1, future=True)
SessionLocal = sessionmaker(autocommit=False, autoflush=False, bind=engine)


def get_db() -> Generator:
    try:
        db = SessionLocal()
        yield db
    finally:
        db.close()


# in my celery task

@celery_app.task()
def some_task():
    db = next(get_db()) # the connection remains after the task is finished
    db.query(.....)
    return 'success'
```

### Description

I used the `next(get_db())` method to use the db session outside the request because the `Depends` cannot be used here as I found out from [#1693](https://github.com/tiangolo/fastapi/issues/1693) but when more tasks get executed the db no longer accepts connection as it had 62 connections already.

I have looked around and found 2 ways people recommend for using the db session
1- `contextmanager`  from [Stackoverflow question](https://stackoverflow.com/questions/66558958/fastapi-dependencies-yield-how-to-call-them-manually)
```python
@celery_app.task()
def some_task():
    with contextmanager(get_db)() as db:
        db.query(.....)
        return 'success'
```

2- another example from [SQLAlchemy docs ](https://docs.sqlalchemy.org/en/14/orm/session_basics.html#when-do-i-construct-a-session-when-do-i-commit-it-and-when-do-i-close-it) is
```python
@celery_app.task()
def some_task():
    with SessionLocal() as db:
        with db.begin():
            db.query(.....)
            return 'success'
```

Both methods significantly reduced the number of connections and close session properly after task while maintaining similar queries/sec performance in a small test I made.
Which of these is recommended as best practice and what are the pros and cons of each method?
how can I optimize the `create_engine` arguments for better connection and less problems with increased queries?

### Operating System

Linux, Windows

### FastAPI Version

0.73.0

### Python Version

3.8

**Comment by dgtlntv** [comment]
I just ran into the same problem as you, and your question put me on the right track to fix it!
Reading through the fastapi ["Dependencies with yield" documentation](https://fastapi.tiangolo.com/tutorial/dependencies/dependencies-with-yield/#using-context-managers-in-dependencies-with-yield) it seems that the preferred way is with contextmanagers, as that is what is used internally in fastapi. I hope I was able to help you with this!

**Comment by Harry2049** [comment]
> I just ran into the same problem as you, and your question put me on the right track to fix it! Reading through the fastapi ["Dependencies with yield" documentation](https://fastapi.tiangolo.com/tutorial/dependencies/dependencies-with-yield/#using-context-managers-in-dependencies-with-yield) it seems that the preferred way is with contextmanagers, as that is what is used internally in fastapi. I hope I was able to help you with this!

Hi! Can you give example of your code?

**Comment by dgtlntv** [comment]
database.py
```python
databaseURL = os.getenv("DATABASE_URL")
engine = create_engine(databaseURL, pool_pre_ping=True, pool_size=32, max_overflow=64)
SessionLocal = sessionmaker(autocommit=False, autoflush=False, bind=engine)
Base = declarative_base()

@contextmanager
def SessionManager():
    db = SessionLocal()
    try:
        yield db
    except:
        # if we fail somehow rollback the connection
        warnings.warn("We somehow failed in a DB operation and auto-rollbacking...")
        db.rollback()
        raise
    finally:
        db.close()
```

somewhere_where_you_need_a_session.py

```python
import database

with database.SessionManager() as db:
    # your code using the database connection goes here
```

**Reply by rcholic** [reply]
got following to work for async_session, just to share with people with same needs as me

**database.py:**
```
from contextlib import asynccontextmanager
from sqlalchemy.ext.asyncio import async_sessionmaker, create_async_engine

engine = create_async_engine(db_url())
async_session = async_sessionmaker(engine, expire_on_commit=False)

@asynccontextmanager
async def AsyncSessionManager():
    db = async_session()
    try:
        yield db
    except Exception as ex:
        print("exception in AsyncSessionManager: ", ex)
        await db.rollback()
        raise
    finally
        await db.aclose()


def db_url() -> URL:
        """
        Assemble database URL from settings.

        :return: database URL.
        """
        return URL.build(
            scheme="postgresql+asyncpg",
            host=self.db_host,
            port=self.db_port,
            user=self.db_user,
            password=self.db_pass,
            path=f"/{self.db_base}",
        )

```
To use async session from database.py:

```
from database import AsyncSessionManager

async with AsyncSessionManager() as async_session:
     # perform database interation with async_session
```

**Comment by yixiongngvsys** [comment]
Thanks @dgtlntv for your solution, you saved my day! I was searching around but still unable to find a solution for db connection that is independent of FastAPI `Depends`. Once again, thanks a lot

**Comment by farahats9** [comment, marked as answer]
here is how I currently do this asynchronously with `SQLAlchemy+asyncpg`

```python
import warnings
from contextlib import asynccontextmanager

from sqlalchemy.ext.asyncio import create_async_engine
from sqlalchemy.orm import sessionmaker
from sqlmodel.ext.asyncio.session import AsyncSession

engine = create_async_engine(...)
SessionLocal = sessionmaker(engine, class_=AsyncSession, expire_on_commit=False, autocommit=False, autoflush=False)

async def get_db() -> AsyncSession:
    async with SessionLocal() as db:
        try:
            yield db
        except:
            # if we fail somehow rollback the connection
            warnings.warn("We somehow failed in a DB operation and auto-rollbacking...")
            await db.rollback()
            raise


@asynccontextmanager
async def get_db_context() -> AsyncSession:
    async with SessionLocal() as db:
        try:
            yield db
        except:
            # if we fail somehow rollback the connection
            warnings.warn("We somehow failed in a DB operation and auto-rollbacking...")
            await db.rollback()
            raise

```
now when using in a request with `Depends` I use `get_db()` but when in a celery task I do this:
```python
async def some_task():
    async with get_db_context() as db:
        ...
```

keep in mind this will increase the number of db session if you run many tasks at the same time in the background since you are creating a new session every time the coroutine is called.
